core/src/test/java/kafka/server/share/SharePartitionManagerTest.java

```diff
@@ -0,0 +1,916 @@
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.server.share.ShareSessionCache;
+import org.apache.kafka.server.share.ShareSessionKey;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.LinkedHashMap;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
+
+@Timeout(120)
+public class SharePartitionManagerTest {
+
+    private static final int RECORD_LOCK_DURATION_MS = 30000;
+    private static final int MAX_DELIVERY_COUNT = 5;
+    private static final short MAX_IN_FLIGHT_MESSAGES = 200;
+
+    private final List<TopicIdPartition> emptyPartList = Collections.unmodifiableList(new ArrayList<>());
+
+    @Test
+    public void testNewContextReturnsFinalContext() {
+        SharePartitionManager sharePartitionManager = SharePartitionManagerBuilder.builder().build();
+
+        ShareFetchMetadata newReqMetadata = new ShareFetchMetadata(Uuid.ZERO_UUID, -1);
+        ShareFetchContext shareFetchContext = sharePartitionManager.newContext("grp", new HashMap<>(), new ArrayList<>(), newReqMetadata);
+        assertEquals(shareFetchContext.getClass(), SharePartitionManager.FinalContext.class);
+
+        // If the final fetch request has topics to add, it should fail as an invalid request
+        Uuid topicId = Uuid.randomUuid();
+        Map<TopicIdPartition, ShareFetchRequest.SharePartitionData> shareFetchData = new HashMap<>();
+        shareFetchData.put(new TopicIdPartition(topicId, new TopicPartition("foo", 0)),
+                new ShareFetchRequest.SharePartitionData(topicId, 4000));
+        assertThrows(InvalidRequestException.class,
+                () -> sharePartitionManager.newContext("grp", shareFetchData, new ArrayList<>(), new ShareFetchMetadata(Uuid.ZERO_UUID, -1)));
+
+        // shareFetchData is not empty, but the maxBytes of topic partition is 0, which means this is added only for acknowledgements.
+        // New context should be created successfully
+        shareFetchData.clear();
+        shareFetchData.put(new TopicIdPartition(topicId, new TopicPartition("foo", 0)),
+                new ShareFetchRequest.SharePartitionData(topicId, 0));
+        shareFetchContext = sharePartitionManager.newContext("grp", shareFetchData, new ArrayList<>(), newReqMetadata);
+        assertEquals(shareFetchContext.getClass(), SharePartitionManager.FinalContext.class);
+    }
+
+    @Test
+    public void testNewContext() {
```

Comment:
Can we divide the test to have different cases i.e. `testNewContextWithFinalContext`, etc.?

Reply:
Yes, I can only segregate out one of the cases from the following test because for others I need them to work one after the other